### PR TITLE
Remove groups

### PR DESCRIPTION
**Pull request by krenzland**
Add script that can remove entire groups (ie a waterlayer) of a mesh. All elements that were neighbors of this group instead have a free surface boundary condition.

WIP but feedback is welcome

### CODE REVIEW

File: remove-groups/src/main.cpp [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+#include "Reader.h"
+#include "Filter.h"
+#include "Writer.h"
+
+int main(int argc, char **argv) {
+  MPI_Init(&argc, &argv);
+  utils::Args args;
+  args.addOption("input", 'i', "input mesh");
+  args.addOption("output", 'o', "output mesh");
+  args.addOption("remove-groups", 'g', "groups to remove");
```
Comment (Thomas-Ulrich) [inline remove-groups/src/main.cpp]: consider precising (as a comma-separated list of id)?

File: remove-groups/src/Filter.h [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+  std::unordered_map<int, int> vertexMap;
+
+  std::cout << "Finding cells to remove" << std::endl;
+  auto elementCounter = 0UL;
+  auto vertexCounter = 0UL;
+  for (auto cell = 0u; cell < mesh.elementSize; ++cell) {
+    if (groupsToRemove.count(mesh.elementGroups[cell]) > 0) {
+      // Remove cell
+      ++removedCells;
+      mesh.elementGroups[cell] = 42;
```
Comment (Thomas-Ulrich) [inline remove-groups/src/Filter.h]: I don't see why you need that.

File: remove-groups/src/Filter.h [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+      const auto &neighbors = mesh.neighbors[cell];
```
Comment (Thomas-Ulrich) [inline remove-groups/src/Filter.h:21]: consider adding following comment:
for each neighbor cell of the current cell, update the component of the boundary array associated with the current cell to 1 (free surface).

File: remove-groups/src/Reader.cpp [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+      connect[connectOffset++] = vertexId;
+    }
+
+    unsigned int faceids[4];
+    PUML::Downward::faces(puml, cells[cell], faceids);
+
+    PUML::Neighbor::face(puml, cell, neighborsRaw.data());
+    neighbors_t curNeighbors{};
+    for (int face = 0; face < 4; ++face) {
+      curNeighbors[face] = (neighborsRaw[face] >= 0) ? std::optional(neighborsRaw[face]) : std::nullopt;
```
Comment (Thomas-Ulrich) [inline remove-groups/src/Reader.cpp:54]: why the std::optional?

File: remove-groups/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,46 @@
+cmake_minimum_required(VERSION 3.17)
+
+project(remove-groups LANGUAGES C CXX)
+
+option(sanitizer "Use address sanitizer when debugging" ON)
```
Comment (Thomas-Ulrich) [inline remove-groups/CMakeLists.txt]: consider using default OFF.

File: remove-groups/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,46 @@
+cmake_minimum_required(VERSION 3.17)
```
Comment (Thomas-Ulrich) [inline remove-groups/CMakeLists.txt]: works with 3.16.5

**Review by Thomas-Ulrich — APPROVED** [review]
Tested on the Samos example. Works.

File: remove-groups/src/Filter.h (hunk 2 above)
Comment (krenzland) [reply remove-groups/src/Filter.h]: Ah, that was just for debugging. I'll remove it :)

**Comment by krenzland** [comment]
Thanks for the review, I'll try to include your feedback next week. There's still some debugging code left that I am going to remove :)

File: remove-groups/src/Reader.cpp (hunk 4 above)
Comment (krenzland) [reply remove-groups/src/Reader.cpp:54]: Boundary cells do not have a neighbor

Comment (Thomas-Ulrich) [reply remove-groups/src/Reader.cpp:54]: so what is the value of neighborsRaw[face] in this case?

**Pull request merged**